eth/tracers/native/erc7562.go

```diff
@@ -0,0 +1,274 @@
+// Copyright 2021 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package native
+
+import (
+	"encoding/json"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/ethereum/go-ethereum/core/tracing"
+	"github.com/ethereum/go-ethereum/core/vm"
+	"github.com/ethereum/go-ethereum/eth/tracers"
+	"github.com/ethereum/go-ethereum/params"
+	"github.com/holiman/uint256"
+	"math/big"
+	"regexp"
+	"strings"
+)
+
+//go:generate go run github.com/fjl/gencodec -type callFrame -field-override callFrameMarshaling -out gen_callframe_json.go
+
+func init() {
+	tracers.DefaultDirectory.Register("callTracerWithOpcodes", newCallTracerWithOpcodes, false)
+}
+
+type callFrameWithOpcodes struct {
+	callFrame
+	AccessedSlots     accessedSlots          `json:"accessedSlots"`
+	ExtCodeAccessInfo []common.Address       `json:"extCodeAccessInfo"`
+	DeployedContracts []common.Address       `json:"deployedContracts"`
+	UsedOpcodes       map[string]bool        `json:"usedOpcodes"`
+	GasObserved       bool                   `json:"gasObserved"`
```

Comment:
This is a new mechanism, trying to detect "gas/call" sequence, which doesn't use the stadnard "delay by 1" opcode check.
There is at least one more case we need such delay mechanism: EXTCODESIZE access to entrypoint is not allowed. But it is OK in the sequence of EXTCODESIZE+ISZERO

Reply:
The point is that the "opcode count" structure ignoring the `GAS` opcode if it was not "observed" is very weird and confusing. Extracting it into a separate field makes much more sense to me.